### PR TITLE
Add Haskell client library

### PR DESCRIPTION
**Pull request by rightfold**
Now we can write the integration tests in Haskell. :+1:

### CODE REVIEW

File: server/integration-tests/listener.html [hunk 1]
```diff
@@ -17,7 +17,6 @@
     console.log('Received data: ' + event.data);
     var status = document.getElementById("status");
     console.log('status: ' + status);
-    status.innerHTML = JSON.stringify(msg);
```
Comment (rkrzr) [inline server/integration-tests/listener.html]: Good catch. This is what happens when I try to write some Javascript.

File: README.md [hunk 2]
```diff
@@ -16,12 +16,13 @@ Do not use it.
...
 Run the tests with `stack test`.
 And run `icepeak` itself with `stack exec icepeak`.
-Install with `stack install`
+Install with `stack install`.
 
-Integration tests are in `/integration-tests`.
+Integration tests are in `/server/integration-tests`.
 They are stand-alone scripts that can be executed directly, e.g. `./connection_test.py`.
 
 ## Usage:
```
Comment (rkrzr) [inline README.md]: Can you add some instructions for the client as well?

File: client-haskell/src/Icepeak/Client.hs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  , clientIcepeakPort :: Word16
+  , clientIcepeakAuth :: ByteString
+  }
+
+-- | Set a value at the leaf of a path.
+--
+-- Will rethrow any exceptions thrown by the I/O actions from
+-- "Network.HTTP.Client". Will not throw any other exceptions.
+setAtLeaf :: (MonadIO m, ToJSON a) => Client -> [Text] -> a -> m HTTP.Status
+setAtLeaf (Client http host port auth) path leaf =
```
Comment (rkrzr) [inline client-haskell/src/Icepeak/Client.hs]: s/http/httpManager

File: client-haskell/src/Icepeak/Client.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    { HTTP.method = "PUT"
+    , HTTP.path = requestPathForIcepeakPath path
+    , HTTP.requestBody = HTTP.RequestBodyLBS (Aeson.encode leaf)
+    }
+
+-- | Return the request path for an Icepeak path.
+requestPathForIcepeakPath :: [Text] -> ByteString
+requestPathForIcepeakPath =
+  ByteString.Lazy.toStrict .
+    Binary.Builder.toLazyByteString .
```
Comment (rkrzr) [inline client-haskell/src/Icepeak/Client.hs]: Is there no way to directly get a strict bytestring from the path?

File: client-haskell/src/Icepeak/Client.hs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+-- | Set a value at the leaf of a path.
+--
+-- Will rethrow any exceptions thrown by the I/O actions from
+-- "Network.HTTP.Client". Will not throw any other exceptions.
+setAtLeaf :: (MonadIO m, ToJSON a) => Client -> [Text] -> a -> m HTTP.Status
+setAtLeaf (Client http host port auth) path leaf =
+  let request = setAtLeafRequest path leaf
+                & (\r -> r { HTTP.host = host, HTTP.port = fromIntegral port })
+                & HTTP.setQueryString [("auth", Just auth)]
```
Comment (rkrzr) [inline client-haskell/src/Icepeak/Client.hs]: Couldn't you set the `host`, the `port` and the query string in the `setAtLeafRequest` function?
Then all the request building code would be in one place.

File: client-haskell/src/Icepeak/Client.hs (hunk 4 above)
Comment (rightfold) [reply client-haskell/src/Icepeak/Client.hs]: Nope. :sob:

**Comment by rightfold** [comment]
> Perhaps you can still add a Haskell integration test that demonstrates how to use the client?

**Pull request merged**